qiskit/algorithms/minimum_eigen_solvers/adapt_qaoa.py

```diff
@@ -0,0 +1,539 @@
+        # Recreates a circuit based on operator parameter.
+        self.ansatz = QAOAAnsatz(
+            cost_operator=operator,
+            initial_state=self._initial_state,
+            mixer_operator=self.optimal_mixer_list,
+            name=self.name,
+        )
+        beta_bounds = self._reps * [(-2 * np.pi, 2 * np.pi)]
+        gamma_bounds = self._reps * [(-np.pi, np.pi)]
+
+        self.ansatz._bounds = beta_bounds + gamma_bounds
+
+    def compute_energy_gradient(self, mixer, operator, ansatz=None) -> ComposedOp:
+        """Computes the energy gradient of the cost operator wrt the mixer pool at an
+            ansatz layer specified by the input 'state' and initial point.
+
+        Returns:
+            The mixer operator with the largest energy gradient along with the
+            associated energy gradient.
+        """
+
+        from qiskit.opflow import commutator
+
+        if not isinstance(operator, MatrixOp):
+            operator = MatrixOp(Operator(operator.to_matrix()))
+        wave_function = ansatz.assign_parameters(self.hyperparameter_dict)
+        # construct expectation operator
+        exp_hc = (self.initial_point[-1] * operator).exp_i()
+        exp_hc_ad = exp_hc.adjoint().to_matrix()
+        exp_hc = exp_hc.to_matrix()
+        energy_grad_op = (
+            exp_hc_ad
+            @ (commutator(operator.to_matrix_op(), mixer.to_matrix_op()).to_matrix())
+            @ exp_hc
+        )
+        energy_grad_op = PrimitiveOp(energy_grad_op)
+
+        expectation = ExpectationFactory.build(
+            operator=energy_grad_op,
+            backend=self.quantum_instance,
+            include_custom=self._include_custom,
+        )
+        observable_meas = expectation.convert(StateFn(energy_grad_op, is_measurement=True))
+        ansatz_circuit_op = CircuitStateFn(wave_function)
+        expect_op = observable_meas.compose(ansatz_circuit_op).reduce()
+        return -1j * expect_op
+
+    def _test_mixer_pool(self, operator: OperatorBase):
+        self._check_problem_configuration(operator)
+        energy_gradients = []
+        for mixer in self.mixer_pool:
+            expect_op = self.compute_energy_gradient(mixer, operator, ansatz=self.ansatz)
+            # run expectation circuit
+            sampled_expect_op = self._circuit_sampler.convert(
+                expect_op, params=self.hyperparameter_dict
+            )
+            meas = sampled_expect_op.eval()
+            energy_gradients.append(meas)
+        max_energy_idx = np.argmax(np.real(energy_gradients))
+        return self.mixer_pool[max_energy_idx], np.abs(energy_gradients[max_energy_idx])
+
+    def compute_minimum_eigenvalue(
+        self,
+        operator: OperatorBase,
+        aux_operators: Optional[List[Optional[OperatorBase]]] = None,
+        iter_results=False,
+        disp=False,
+    ):
+        self.optimal_mixer_list = None
+        """Runs ADAPT-QAOA for each iteration"""
```

Comment:
Shouldn't `optimal_mixer_list` be a list? Because on L235 you use the `append` method 🙂 

Also, this method misses type docs.
```suggestion
        """Runs ADAPT-QAOA for each iteration"""
        self.optimal_mixer_list = []
```